<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h1 id="creating-reference-indices">Creating reference indices</h1>
<p>The <a href="usage-instructions.html">usage instructions</a> give the necessary commands to create the reference indicies required for Rp-Bp. This document describes how to download or otherwise find the necessary files and sequences.</p>
<h2 id="reference-genome-sequence">Reference genome sequence</h2>
<p>While the most recent version available of the reference genome sequence should be used, we do not recommend to include things like haplotypes in the reference for Rp-Bp. First, Rp-Bp does not treat these differently than “normal” chromosomes. Second, the inclusion of these can greatly increase the alignment step (with STAR). Most importantly, though, if these extra sequences substantially overlap the main reference, reads which align to both will be called as multimappers and discarded.</p>
<p>As mentioned in the <a href="usage-instructions.html">usage instructions</a>, the “primary assembly” file from Ensembl contains the appropriate sequences and identifiers. The “top level” Ensembl genome assembly includes haplotype information. Please see <a href="http://www.ensembl.org/info/genome/genebuild/assembly.html">Ensembl</a> for more information about the differences between assemblies.</p>
<h2 id="reference-annotations">Reference annotations</h2>
<p>Again, we recommend using the most recent version of annotations available. The annotations <strong>must match</strong> the version of the reference genome sequence. As described in the <a href="usage-instructions.html">usage instructions</a>, at least the <code>exon</code> features must be present, and the transcript identifiers (<code>transcript_id</code> attribute for GTF) must match for exons from the same transcript. The <code>CDS</code> features must also be present in the annotation file. Rp-Bp’s ORF labeling algorithm assumes that annotated CDSs <em>include</em> the START codon but <em>exclude</em> the STOP codon.</p>
<p>There seem to be no clear consensus if the STOP codon should be part of the coding region. The GTF2 (GTF) specifications suggest to exclude it, and this is the default in Rp-Bp when the file as a <em>.gtf</em> extension. If the annotations follow the GFF3 specifications (file has <em>.gff</em> extension), the STOP codon <em>is included</em> and will be automatically removed during processing. This can be verified by loading the annotations and sequence into a genome viewer such as IGV. If there is a problem with the annotations, it is likely to manifest as many <code>within</code> and <code>suspect</code> ORFs while very few <code>canonical</code> ORFs predicted as translated.</p>
<p>Following standard conventions, the genomic locations are taken to be base-1 and inclusive. Other feature types, such as <code>gene</code>, <code>start_codon</code> and <code>stop_codon</code> may be present in the file, but they <strong>are not used</strong> for extracting ORFs.</p>
<p><strong>N.B.</strong> In case CDS annotations are <em>not</em> available, Rp-Bp can still be run as normal. The only difference is that the ORF labels will not be meaningful.</p>
<h2 id="ribosomal-index">“ribosomal” index</h2>
<p>The ribosomal filtering step is not as imperative for Rp-Bp as for some other types of analysis (such as anything based on RPKM); this is because reads which align to these types of sequences tend to be multimappers and are discarded anyway. Nevertheless, filtering as many of these as possible ensures the rest of the pipeline runs more efficiently.</p>
<p>Further, the ribosomal subunits, and other similar sequences, are often not included in genome annotations. Thus, it is possible that reads which map to the ribosomal subunit end up mapping uniquely to the genome. Thus, we recommend a comprehensive ribosomal index for filtering.</p>
<p>We typically include the following in our indices:</p>
<ul>
<li><p>The large and small ribosomal subunit sequences, typically from NCBI. These can be surprisingly difficult to track down; nevertheless, a search using the terms: <code>&quot;ribosomal subunit&quot;</code> and <code>&lt;organism&gt;</code> (including the quotes around <code>ribosomal subunit</code>) typically works. (<a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=%22ribosomal+subunit%22+mouse">example</a>)</p></li>
<li><p>The genomic tRNA sequences from <a href="http://gtrnadb.ucsc.edu/">GtRNAdb</a>. Since only the sequences are used, it is not very important if the annotation version in GtRNAdb does not match that used in the rest of the analysis. For example, using the tRNAs from GRCh37 and annotations from GRCh38 should not cause problems. Of course, if matching versions are available, it is preferred to use them.</p></li>
<li><p>“Mt_rRNA”, “Mt_tRNA” and “rRNA” genes from BioMart. In particular, we select those options for the “Gene type” filter. For “Attributes”, we select “Sequences”, and then specifically “Exon sequences”. Additionally, including the “Gene type” in the header can be helpful later for identifying where reads in the <code>with-rrna</code> folder mapped.</p></li>
</ul>
<p>Obviously, depending on the goal of the analysis, the choice of indicies could vary. For example, it may be desirable to include tRNA in the analysis or to exclude known non-coding RNA species, such as snRNAs. These could also be included in the “Gene type”s from BioMart.</p>
<h2 id="de-novo-assembly"><em>de novo</em> assembly</h2>
<p>Often, matching RNA-seq is available for riboseq datasets. In these cases, we highly recommend creating a <em>de novo</em> assembly from the RNA-seq data. A variety of algorithms are available for this task. Internally, we use a custom pipeline which is based on the STAR aligner and StringTie assembler. Many, many other options exist for transcript assembly, though. The only requirement for use with Rp-Bp is that the assembler produces a valid GTF/GFF file (or something which can be converted into GTF/GFF). The coordinates in the GTF/GFF file must match those in the reference genome.</p>
<p><em>de novo</em> assembly algorithms do not typically identify coding regions (that is what riboseq is for!); however, if they do include CDS annotations, those should match the same start and stop codon conventions described above.</p>
</body>
</html>
