<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h1 id="running-the-rp-bp-pipeline">Running the Rp-Bp pipeline</h1>
<p>The Rp-Bp pipeline consists of an index creation step, which must be performed once for each genome and set of annotations, and a two-phase prediction pipeline, which must be performed for each sample.</p>
<p>In the first phase of the prediction pipeline, a filtered genome profile is created. In the second phase, the ORFs which show evidence of translation in the profile are predicted.</p>
<p>This document describes the steps to run each of these phases. It shows some sample calls. For all programs, the <code>--help</code> option can be given to see the complete list of parameters.</p>
<p><a id="toc"></a></p>
<ul>
<li><a href="#creating-reference-genome-indices">Creating reference genome indices</a></li>
<li><a href="#running-pipelines">Running the Rp-Bp pipeline</a>
<ul>
<li><a href="#creating-filtered-genome-profiles">Creating filtered genome profiles</a></li>
<li><a href="#predicting-translated-open-reading-frames">Predicting translated open reading frames</a></li>
<li><a href="#using-existing-alignment-files">Using existing alignment files</a></li>
</ul></li>
<li><a href="#logging-options">Logging options</a></li>
<li><a href="#parallel-processing-options">Parallel processing options</a> (SLURM options)</li>
</ul>
<p><a id="creating-reference-genome-indices"></a></p>
<h2 id="creating-reference-genome-indices">Creating reference genome indices</h2>
<p>This section describes the steps necessary to prepare a reference genome and matching annotation for use in the Rp-Bp pipeline. The process must only be run once for each reference genome and set of annotations.</p>
<p>Please see <a href="creating-reference-indices.html">creating reference indices</a> for more details about choosing the appropriate files from Ensemble, etc., for Rp-Bp.</p>
<p>The entire index creation process can be run automatically using the <code>prepare-rpbp-genome</code> scripts. It reads most of the required paths from a <a href="http://www.yaml.org/start.html">YAML</a> configuration file. Additionally, it automatically creates some of the output paths.</p>
<p>The script accepts a <code>--overwrite</code> flag. Unless this is given, then steps for which the output file already exists will be skipped.</p>
<p>It also accepts a <code>--do-not-call</code> flag. If this flag is given, then the commands below will be printed but not executed.</p>
<p><a href="#logging-options">Logging options</a> can be given to this script.</p>
<p><a href="#parallel-processing-options">Parallel processing options</a> can be given to this script.</p>
<h3 id="configuration-file-keys">Configuration file keys</h3>
<p>The following keys are read from the configuration file. Keys with [<code>brackets</code>] are optional.</p>
<ul>
<li><code>gtf</code>. The path to the reference annotations</li>
<li><code>fasta</code>. The path to the reference genome sequence</li>
<li><p><code>ribosomal_fasta</code>. The path to the ribosomal sequence</p></li>
<li><p>[<code>de_novo_gtf</code>]. An additional GTF/GFF containing annotations constructed from a <em>de novo</em> assembly (for example, from StringTie). See below for how <em>de novo</em> assemblies are handled. This file and the reference annotation file must have the same format (both GTF or GFF3), see below regarding format specifications.</p></li>
<li><code>genome_base_path</code>. The path to the output directory for the transcript fasta and ORFs</li>
<li><code>genome_name</code>. A descriptive name to use for the created files</li>
<li><code>ribosomal_index</code>. The base output path for the Bowtie2 index of the ribosomal sequence</li>
<li><p><code>star_index</code>. The base output path for the STAR index for the reference genome sequence. This STAR index <em>does not</em> include the transcript annotations. They will be incorporated later while running the specific samples.</p></li>
<li>[<code>orf_note</code>]. An additional description used in the filename of the created ORFs</li>
<li>[<code>start_codons</code>]. A list of strings that will be treated as start codons when searching for ORFs. default: [<code>ATG</code>]</li>
<li>[<code>stop_codons</code>]. A list of strings that will be treated as stop codons when searching for ORFS. default: [<code>TAA</code>, <code>TGA</code>, <code>TAG</code>]</li>
<li><p>[<code>sjdb_overhang</code>]. The value to use for splice junction overlaps when constructing the STAR index. default: 50</p></li>
</ul>
<h3 id="orf-labels">ORF labels</h3>
<p>As part of the index creation phase, ORFs are labeled according to their location relative to the annotated, canonical transcripts, that is, annotated CDS regions. We use the following labels (essentially, the same as those given in the supplement of the paper):</p>
<ul>
<li><code>canonical</code>. an ORF which exactly coincides with an annotated transcript</li>
<li><code>canonical_extended</code>. an ORF which starts upstream of a <code>canonical</code> ORF, but otherwise has the same structure</li>
<li><code>canonical_truncated</code>. an ORF which starts downstream of a <code>canonical</code> ORF, but otherwise has the same structure</li>
<li><code>five_prime</code>. an ORF which is completely in the annotated 5’ leader region of a transcript and does not overlap other <code>canonical</code> ORFs on the same strand (such as from another isoform)</li>
<li><code>three_prime</code>. an ORF in the annotated 3’ trailer region of a transcript and does not overlap other <code>canonical</code> ORFs on the same strand</li>
<li><code>noncoding</code>. an ORF from a transcript not annotated as coding, such as a lncRNA or pseudogene.</li>
<li><code>five_prime_overlap</code>. an ORF in the annotated 5’ leader region of a transcript but which overlaps a <code>canonical</code> ORF; this could either be out-of-frame with respect to the <code>canonical</code> ORF of the same transcript or otherwise overlap a <code>canonical</code> ORF from another isoform.</li>
<li><code>three_prime_overlap</code>. an ORF in the annotated 3’ trailer region of a transcript but which overlaps a <code>canonical</code> ORF</li>
<li><code>suspect</code>. an ORF which partially overlaps the interior of a <code>canonical</code> ORF. These can result from things like retained introns.</li>
<li><code>within</code>. an out-of-frame ORF in the interior of a <code>canonical</code> ORF</li>
</ul>
<h3 id="input-files">Input files</h3>
<p>The required input files are those suggested by the configuration file keys.</p>
<ul>
<li><p><code>gtf</code>. The GTF/GFF file containing the reference annotations. The file must follow standard conventions for annotating transcripts. This means at least the <code>exon</code> features must be present, and the transcript identifiers (<code>transcript_id</code> attribute for GTF) must match for exons from the same transcript. Furthermore, the ORFs are labeled based on their positions relative to annotated coding sequences. For it to work correctly, the <code>CDS</code> features must also be present in the annotation file. The START codon should be included in the <code>CDS</code>, but the STOP codon should not. By default, Rp-Bp treats annotations according to the GTF2 (GTF) specifications, when the file extension is <em>.gtf</em>. Limited support is provided for <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.html">GFF3 specifications</a>. If the file extension is <em>.gff</em>, it will be treated as GFF3. In this case, it is assumed that the START and STOP codons are both included in the CDS. Rp-Bp will then automatically remove the STOP codon during processing. Following standard conventions, the genomic locations are taken to be base-1 and inclusive. Other feature types, such as <code>gene</code>, <code>start_codon</code> and <code>stop_codon</code> may be present in the file, but they <strong>are not used</strong> for extracting ORFs.</p></li>
<li><p><code>fasta</code>. The input fasta file should contain all chromosome sequences. The identifiers must match those in the GTF file. Typically, the “primary assembly” file from Ensembl contains the appropriate sequences and identifiers. In particular, the “top level” Ensembl genome assembly includes haplotype information that is not used by Rp-Bp; for human, this is quite large and can substantially slow down the pipeline. Please see <a href="http://www.ensembl.org/info/genome/genebuild/assembly.html">Ensembl</a> for more information about the differences between assemblies.</p></li>
<li><p><code>ribosomal_fasta</code>. The ribosomal DNA sequence is typically repeated many times throughout the genome. Consequently, it can be difficult to include in the genome assembly and is often omitted. Therefore, a separate fasta file is required for these sequences (which are later used to filter reads). This file could also include other sequences which should be filtered, such as tRNA sequences downloaded from <a href="http://gtrnadb.ucsc.edu/">GtRNAdb</a>, for example.</p></li>
<li><p><code>de_novo_gtf</code>. The GTF/GFF file containing the <em>de novo</em> assembled transcripts. It must have the same format specifications at the <code>gtf</code> file.</p></li>
</ul>
<h3 id="de-novo-assembled-transcripts"><em>de novo</em> assembled transcripts</h3>
<p>In principle, there is no difference between “standard” annotated transcripts and <em>de novo</em> assembled transcripts. In both cases, ORFs are extracted from the transcripts based on the given <code>start_codons</code> and <code>stop_codons</code>. However, it is often of scientific interest to distinguish between ORFs from annotated transcripts and “novel” ones from the <em>de novo</em> assembly.</p>
<p>Both to avoid reporting spurious “novel” translated ORFs and to avoid redundant calculations, we remove each ORF identified in the <em>de novo</em> assembly which exactly match an ORF using only the annotations.</p>
<p>We then label ORFs from the <em>de novo</em> assembly as follows, with respect to their relationship to the annotated transcripts. Due to the first filtering step, the following categories include only ORFs which are at least partially only supported in the <em>de novo</em> assembly.</p>
<ul>
<li><code>novel</code>. The ORF does not overlap the annotations at all.</li>
<li><code>novel_canonical_extended</code>. an ORF which starts upstream of a <code>canonical</code> ORF, but otherwise has the same structure.</li>
<li><code>novel_noncoding</code>. an ORF from a transcript not annotated as coding, such as a lncRNA or pseudogene.</li>
<li><code>novel_five_prime_overlap</code>. an ORF in the annotated 5’ leader region of a transcript but which overlaps a <code>canonical</code> ORF; this could either be out-of-frame with respect to the <code>canonical</code> ORF of the same transcript or otherwise overlap a <code>canonical</code> ORF from another isoform.</li>
<li><code>novel_three_prime_overlap</code>. an ORF in the annotated 3’ trailer region of a transcript but which overlaps a <code>canonical</code> ORF</li>
<li><code>novel_suspect</code>. an ORF which partially overlaps the interior of a <code>canonical</code> ORF. These can result from things like retained introns.</li>
<li><code>novel_within</code>. an out-of-frame ORF in the interior of a <code>canonical</code> ORF</li>
</ul>
<p>The other labels, such as <code>novel_canonical</code> or <code>novel_five_prime</code> are guaranteed to be filtered in the first step based on their definitions.</p>
<p>In preparation to running the main pipeline, <code>reference</code> and <code>de_novo</code> annotations are concatenated in a single file which is used on the fly by <code>STAR</code> at the mapping step. Merging and sorting the resulting annotations is not necessary.</p>
<h3 id="output-files">Output files</h3>
<ul>
<li><p><code>&lt;genome_base_path&gt;/ribosomal_index/</code>. The bowtie2 index files (<code>ribosomal_index</code>.1.bt2, etc.) for the ribosomal_fasta file</p></li>
<li><p><code>&lt;genome_base_path&gt;/star_index/</code>. The STAR index files (<code>SA</code>, <code>Genome</code>, etc.) for the <code>fasta</code> file</p></li>
<li><p><code>&lt;genome_base_path&gt;/&lt;genome_name&gt;.annotated.bed.gz</code>. A bed12 file containing all transcripts. For coding transcripts, the thick_start and thick_end columns give the start and end of the coding region; the start codon <em>is</em> included in the thick region, but the stop codon <em>is not</em>. For noncoding transcripts, thick_start and thick_end are both -1. This seems to behave as expected in IGV.</p></li>
</ul>
<h4 id="from-annotations">From annotations</h4>
<ul>
<li><p><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.transcripts.annotated.fa</code>. The sequences of all annotated transcripts.</p></li>
<li><p><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.genomic-orfs.annotated.&lt;orf_note&gt;.bed.gz</code>. A bed12+ file containing all ORFs from annotated transcripts. Besides the standard bed12 columns, this file includes columns giving the orf_type, orf_length, and orf_num. The ORF ids are of the form: <code>transcript_seqname:start-end:strand</code>. The start codon <em>is</em> included in the ORF, but the stop codon <em>is not</em>. The thick_start and thick_end are always the same as start and end.</p></li>
<li><p><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.orfs.annotated.&lt;orf_note&gt;.bed.gz</code>. A bed6+2 file containing each exon from each ORF. The “id” of an exon corresponds exactly to the ORF to which it belongs; exons from the same ORF have the same “id”. The extra columns are “exon_index”, which gives the order of the exon in the transcript, and “transcript_start”, which gives the start position of that index in transcript coordinates. Exons are always sorted by “lowest start first”, so the order is really reversed for ORFs on the reverse strand.</p></li>
</ul>
<h4 id="from-de-novo-assembly.">From <em>de novo</em> assembly.</h4>
<p>The semantics of these files are the same of those from the annotations, but created using the <code>de_novo_gtf</code> files. N.B., ORFs which completely overlap annotations are not included.</p>
<ul>
<li><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.transcripts.de-novo.fa</code></li>
<li><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.genomic-orfs.de-novo.&lt;orf_note&gt;.bed.gz</code></li>
<li><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.orfs.de-novo.&lt;orf_note&gt;.bed.gz</code></li>
</ul>
<h4 id="from-both-annotations-and-de-novo-assembly.">From both annotations and <em>de novo</em> assembly.</h4>
<p>The semantics are again the same as above. If a <em>de novo</em> assembly was not provided, these are simply symlinks to the respective “annotations” files. Otherwise, they are the concatenation of the respective “annotation” and “<em>de novo</em>” files.</p>
<ul>
<li><p><code>&lt;genome_base_path&gt;/&lt;genome_name&gt;.star-input.gtf</code></p></li>
<li><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.genomic-orfs.&lt;orf_note&gt;.bed.gz</code></li>
<li><p><code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.orfs.&lt;orf_note&gt;.bed.gz</code></p></li>
</ul>
<pre><code>prepare-rpbp-genome WBcel235.79.chrI.yaml --num-cpus 2 --mem 4G --overwrite --logging-level INFO</code></pre>
<p><a href="#toc">Back to top</a></p>
<p><a id='running-pipelines'></a></p>
<h2 id="running-the-rp-bp-pipeline-1">Running the Rp-Bp pipeline</h2>
<p>The entire Rp-Bp pipeline can be run on a set of riboseq samples which all use the same genome indices using a sample sheet-like <a href="http://www.yaml.org/start.html">YAML</a> configuration file with the <code>run-all-rpbp-instances</code> program. The configuration file respects all of the optional configuration options specified below.</p>
<p>The script accepts the following options:</p>
<ul>
<li><p><code>--tmp &lt;loc&gt;</code>. If this flag is given, then all relevant calls will use <code>&lt;loc&gt;</code> as the base temporary directory. Otherwise, the program defaults will be used.</p></li>
<li><p><code>--overwrite</code>. Unless this is given, then steps for which the output file already exists will be skipped.</p></li>
<li><p><code>--keep-intermediate-files</code>. Unless this flag is given, large intermediate files, such as fastq files output by flexbar after removing adapters, will be deleted.</p></li>
<li><p><code>--flexbar-format-option</code>. Older versions of flexbar used <code>format</code> as the command line option to specify the format of the fastq quality scores, while newer versions use <code>qtrim-format</code>. Depending on the installed version of flexbar, this option may need to be changed. Default: <code>qtrim-format</code></p></li>
<li><p><code>--star-executable</code>. In principle, <code>STARlong</code> (as opposed to <code>STAR</code>) could be used for alignment. Given the nature of riboseq reads (that is, short due to the experimental protocols of degrading everything not protected by a ribosome), this is unlikely to be a good choice, though. Default: <code>STAR</code></p></li>
<li><p><code>--star-read-files-command</code>. The input for <code>STAR</code> will always be a gzipped fastq file. <code>STAR</code> needs the system command which means “read a gzipped text file”. As discovered in <a href="https://github.com/dieterich-lab/rp-bp/issues/35">Issue #35</a>, the name of this command is different on OSX and ubuntu. The program now attempts to guess the name of this command based on the system operating system, but it can be explicitly specified as a command line option. Default: <code>gzcat</code> if <code>sys.platform.startswith(&quot;darwin&quot;)</code>; <code>zcat</code> otherwise. Please see <a href="https://docs.python.org/3/library/sys.html">python.sys documentation</a> for more details about attempting to guess the operating system.</p></li>
</ul>
<p><a href="#logging-options">Logging options</a> can be given to this script.</p>
<p><a href="#parallel-processing-options">Parallel processing options</a> can be given to this script.</p>
<h3 id="using-replicates">Using replicates</h3>
<p>The Rp-Bp pipeline handles replicates by adding the (smoothed) ORF profiles. The Bayes factors and predictions are then calculated based on the combined profiles. The <code>--merge-replicates</code> flag indicates that the replicates should be merged. By default, if the <code>--merge-replicates</code> flag is given, then predictions will not be made for the individual datasets. The <code>--run-replicates</code> flag can be given to override this and make predictions for both the merged replicates as well as the individual datasets.</p>
<h3 id="configuration-file-keys-1">Configuration file keys</h3>
<p>The following keys are read from the configuration file. Their semantics is exactly as described below. (This text is in both places for convenience.)</p>
<ul>
<li><code>riboseq_data</code>. The base output location for all created files.</li>
<li>[<code>note</code>]. An optional string which will be added to all filenames. It should not contain spaces or any other special characters.</li>
<li>[<code>models_base</code>]. The base path to the compiled models, the base path to the compiled models. The models specified in the paper are included with the source distribution and compiled/pickled as part of the installation process. They are installed in an operating system-specific location (in particular, <code>user_data_dir</code> from the <a href="https://pypi.python.org/pypi/appdirs">appdirs package</a>. This location is determined during installation and does not normally need to be changed. For development, they may be in some alternative location.</li>
</ul>
<h4 id="reference-genome-options">Reference genome options</h4>
<p>These options should be exactly the same as those used in the configuration file used to create the reference indices.</p>
<ul>
<li><code>gtf</code>. The path to the reference annotations</li>
<li><code>genome_base_path</code>. The path to the output directory for the transcript fasta and ORFs</li>
<li><code>genome_name</code>. A descriptive name to use for the created files</li>
<li><code>ribosomal_index</code>. The base path for the Bowtie2 index of the ribosomal sequence</li>
<li><code>star_index</code>. The base path to the STAR index</li>
<li><code>fasta</code>. The path to the reference genome sequence</li>
</ul>
<h4 id="samples-specification">Samples specification</h4>
<ul>
<li><p><code>riboseq_samples</code>. A dictionary in which each entry specifies a sample. The key is an informative name about the sample, and the value gives the complete path to the sequencing file (a fastq or fastq.gz file). The names will be used to construct filenames, so they should not contain spaces or other special characters.</p></li>
<li><p><code>riboseq_biological_replicates</code>. A dictionary in which each entry species one condition and all samples which are replicates of the condition. The key of the dictionary is a string description of the condition, and the value is a list that gives all of the sample replicates which belong to that condition. The names of the sample replicates must match the dataset names specified in <code>riboseq_samples</code>.</p></li>
</ul>
<pre><code># do not merge replicates
run-all-rpbp-instances c-elegans-test.yaml --tmp /scratch/bmalone/ --num-cpus 10 --logging-level INFO --keep-intermediate-files

# merging the replicates, do not calculate Bayes factors and make predictions for individual datasets
run-all-rpbp-instances c-elegans-test.yaml --tmp /scratch/bmalone/ --overwrite --num-cpus 2 --logging-level INFO --merge-replicates --keep-intermediate-files

# merging the replicates and calculating Bayes factors and making predictions for individual datasets
run-all-rpbp-instances c-elegans-test.yaml --tmp /scratch/bmalone/ --overwrite --num-cpus 2 --logging-level INFO --merge-replicates --run-replicates --keep-intermediate-files</code></pre>
<p><a href="#toc">Back to top</a></p>
<p><a id='creating-filtered-genome-profiles'></a></p>
<h2 id="creating-orf-profiles">Creating ORF profiles</h2>
<p>The entire profile creation process can be run automatically using the <code>create-orf-profiles</code> script. It reads most of the required paths from a <a href="http://www.yaml.org/start.html">YAML</a> configuration file. Additionally, it automatically creates some of the output paths.</p>
<p>The script accepts a <code>--overwrite</code> flag. Unless this is given, then steps for which the output file already exists will be skipped.</p>
<p>It also accepts a <code>--do-not-call</code> flag. If this flag is given, then the commands below will be printed but not executed.</p>
<p>The script accepts a <code>--keep-intermediate-files</code> flag. Unless this flag is given, large intermediate files, such as fastq files output by flexbar after removing adapters, will be deleted.</p>
<p><a href="#logging-options">Logging options</a> can also be given to this script.</p>
<p><a href="#parallel-processing-options">Parallel processing options</a> can be given to this script.</p>
<h3 id="configuration-file-keys-2">Configuration file keys</h3>
<p>The following keys are read from the configuration file. Keys with [<code>brackets</code>] are optional.</p>
<ul>
<li><code>riboseq_data</code>. The base output location for all created files.</li>
<li>[<code>note</code>]. An optional string which will be added to all filenames. It should not contain spaces or any other special characters.</li>
<li>[<code>models_base</code>]. The base path to the compiled models, the base path to the compiled models. The models specified in the paper are included with the source distribution and compiled/pickled as part of the installation process. They are installed in an operating system-specific location (in particular, <code>user_data_dir</code> from the <a href="https://pypi.python.org/pypi/appdirs">appdirs package</a>. This location is determined during installation and does not normally need to be changed. For development, they may be in some alternative location.</li>
</ul>
<h4 id="reference-genome-options-1">Reference genome options</h4>
<p>These options should be exactly the same as those used in the configuration file used to create the reference indices.</p>
<ul>
<li><code>gtf</code>. The path to the reference annotations</li>
<li><code>genome_base_path</code>. The path to the output directory for the transcript fasta and ORFs</li>
<li><code>genome_name</code>. A descriptive name to use for the created files</li>
<li><code>ribosomal_index</code>. The base path for the Bowtie2 index of the ribosomal sequence</li>
<li><code>star_index</code>. The base path to the STAR index</li>
</ul>
<h4 id="flexbar-options">Flexbar options</h4>
<ul>
<li>[<code>adapter_file</code>]. A fasta file containing a set of adapter sequences used by flexbar. default: None</li>
<li>[<code>max_uncalled</code>]. The maximum number of Ns to permit in a read without filtering. default 1</li>
<li>[<code>pre_trim_left</code>]. The number of bases to remove from the 5’ end of all reads: default: 0</li>
<li>[<code>adapter_sequence</code>]. A single sequence used to remove adapters within flexbar. default: None</li>
<li>[<code>quality_format</code>]. The quality format of the reads in the raw fastq file. default: “sanger”</li>
</ul>
<h4 id="star-options">STAR options</h4>
<ul>
<li>[<code>align_intron_min</code>]. default: 20</li>
<li>[<code>align_intron_max</code>]. default: 100000</li>
<li>[<code>out_filter_intron_motifs</code>]. default: RemoveNoncanonicalUnannotated</li>
<li>[<code>out_filter_mismatch_n_max</code>]. default: 1</li>
<li>[<code>out_filter_mismatch_n_over_l_max</code>]. default: 0.04</li>
<li>[<code>out_filter_type</code>]. default: BySJout</li>
<li>[<code>out_sam_attributes</code>]. default: AS NH HI nM MD</li>
<li>[<code>sjdb_overhang</code>]. default: 50</li>
</ul>
<h4 id="multimapper-options">Multimapper options</h4>
<ul>
<li>[<code>keep_riboseq_multimappers</code>]: If this variable is present in the config file with any value (even something like “no” or “null” or “false”), then multimapping riboseq reads <em>will not</em> be removed. They will be treated as “normal” reads in every place they map. That is, the weight of the read will not be distributed fractionally, probabilistically, etc.</li>
</ul>
<h4 id="metagene-periodicity-options">Metagene periodicity options</h4>
<ul>
<li><p>[<code>seqids_to_keep</code>]. If this list is given, then only transcripts appearing on these identifiers will be used to construct the metagene profiles (and other downstream analysis). The identifiers must match exactly (e.g., “2” and “chr2” do not match)</p></li>
<li><p>[<code>metagene_profile_start_upstream</code>]. The number of bases upstream of the translation initiation site to begin constructing the metagene profile. default: 50</p></li>
<li><p>[<code>metagene_profile_start_downstream</code>]. The number of bases downstream of the translation initiation site to end the metagene profile. default: 20</p></li>
<li><p>[<code>metagene_profile_end_upstream</code>]. The number of bases upstream of the translation termination site to begin constructing the metagene profile. default: 50</p></li>
<li><p>[<code>metagene_profile_end_downstream</code>]. The number of bases downstream of the translation termination site to end the metagene profile. default: 20</p></li>
<li>[<code>periodic_offset_start</code>]. The position, relative to the translation initiation site, to begin calculating periodicity Bayes factors (inclusive)</li>
<li>[<code>periodic_offset_end</code>]. The position, relative to the translation initiation site, to stop calculating periodicity Bayes factors (inclusive)</li>
<li><p>[<code>metagene_profile_length</code>]. The length of the profile to use in the models. <code>metagene_profile_length</code> + <code>periodic_offset_end</code> must be consistent with the length of the extracted metagene profile</p></li>
<li><p>[<code>metagene_profile_iterations</code>]. The number of iterations to use for each chain in the MCMC sampling. The first half of the iterations are discarded as burn-in samples. All of the remaining samples are used to estimate the posterior distributions. That is, we do not use thinning. default: 500</p></li>
</ul>
<h4 id="periodicity-and-offset-options">Periodicity and offset options</h4>
<ul>
<li><p>[<code>use_fixed_lengths</code>]. If this variable is present in the config file with any value (even something like “no” or “null” or “false”), then the estimated periodic read lengths and offsets will not be used. Instead, fixed values given by <code>lengths</code> and <code>offsets</code> (below) will be used.</p></li>
<li><p>[<code>lengths</code>]. A list of read lengths which will be used for creating the profiles if the <code>use_fixed_lengths</code> option is given. Presumably, these are lengths that have periodic metagene profiles.</p></li>
<li><p>[<code>offsets</code>]. The P-site offset to use for each read length specifed by <code>--lengths</code> if the <code>use_fixed_lengths</code> option is given. The number of offsets must match the number of lengths, and they are assumed to match. For example <code>lengths</code> of 26, 29 with <code>offsets</code> 9, 12 means only reads of lengths 26 bp and 29 bp will be used to create the profiles. The 26 bp reads will be shifted by 9 bp in the 5’ direction, while reads of length 29 bp will be shifted by 12 bp.</p></li>
<li><p>[<code>min_metagene_profile_count</code>]. If fixed lengths are not used: the minimum number of reads for a particular length in the filtered genome profile. Read lengths with fewer than this number of reads will not be used. default: 1000</p></li>
<li><p>[<code>min_metagene_bf_mean</code>]. If fixed lengths are not used: if <code>max_metagene_profile_bayes_factor_var</code> is not None, then this is taken as a hard threshold on the estimated Bayes factor mean.</p></li>
</ul>
<p>If <code>min_metagene_profile_bayes_factor_likelihood</code> is given, then this is taken as the boundary value; that is, a profile is “periodic” if:</p>
<pre><code>   [P(bf &gt; min_metagene_bf_mean)] &gt; min_metagene_bf_likelihood</code></pre>
<p>If both <code>max_metagene_bf_var</code> and <code>min_metagene_bf_likelihood</code> are <code>None</code>, then this is taken as a hard threshold on the mean for selecting periodic read lengths.</p>
<p>If both <code>max_metagene_bf_var</code> and <code>min_metagene_bf_likelihood</code> are given, then both filters will be applied and the result will be the intersection.</p>
<p>default: 5</p>
<ul>
<li><p>[<code>max_metagene_bf_var</code>]. If fixed lengths are not used: if given, then this is taken as a hard threshold on the estimated Bayes factor variance. default: None, i.e., this filter is not used. (null in yaml)</p></li>
<li><p>[<code>min_metagene_bf_likelihood</code>]: If fixed lengths are not used: if given, then this is taken a threshold on the likelihood of periodicity (see <code>min_metagene_bf_mean</code> description for more details). default: 0.5</p></li>
</ul>
<h4 id="smoothing-options">Smoothing options</h4>
<ul>
<li><p>[<code>smoothing_fraction</code>]. The fraction of the profile to use for smoothing within LOWESS. default: 0.2</p></li>
<li><p>[<code>smoothing_reweighting_iterations</code>]. The number of reweighting iterations to use within LOWESS. Please see the statsmodels documentation for a detailed description of this parameter. default: 0</p></li>
<li><p>[<code>min_orf_length</code>]. If this value is greater than 0, then ORFs whose length (in nucleotides) is less than this value will not be smoothed, and neither the Bayes factor estimates nor the chi-square p-value will be calcualted. default: 0</p></li>
<li><p>[<code>max_orf_length</code>]. If this value is greater than 0, then ORFs whose length (in nucleotides) is greater than this value will not be smoothed, and neither the Bayes factor estimates nor the chi-square p-value will be calcualted. default: 0</p></li>
<li><p>[<code>min_signal</code>]. ORFs for which the number of in-frame reads is less than this value will not be smoothed, and neither the Bayes factor estimates nor the chi-square p-value will be calculated. default: 5</p></li>
</ul>
<h4 id="shared-mcmc-options">Shared MCMC options</h4>
<p>These affect the MCMC both for estimating metagene profile periodicity and ORF translation Bayes factors.</p>
<ul>
<li><p>[<code>seed</code>]. The random seed for the MCMC sampling. default: 8675309</p></li>
<li><p>[<code>chains</code>]. The number of chains to use in the MCMC sampling. default: 2</p></li>
</ul>
<h3 id="input-files-1">Input files</h3>
<p>The required input files are only those suggested by the configuration file keys.</p>
<ul>
<li><code>ribosomal_index</code>.</li>
<li>The STAR index.</li>
<li>The periodic models, taken as all <code>.pkl</code> files located in <code>&lt;models_base&gt;/periodic</code>.</li>
<li>The nonperiodic models, taken as all <code>.pkl</code> files located in <code>&lt;models_base&gt;/nonperiodic</code>.</li>
</ul>
<p>In addition, the input annotations <code>&lt;genome_base_path&gt;/&lt;genome_name&gt;.star-input.gtf</code> are used by <code>STAR</code>. For details, see <a href="#creating-reference-genome-indices">Creating reference genome indices</a>.</p>
<h3 id="output-files-1">Output files</h3>
<p>This script primarily creates the following files. (STAR also creates some temporary and log files.)</p>
<p><strong>N.B.</strong> If the <code>keep_riboseq_multimappers</code> configuration option is given, then the “-unique” part will not be present in the output filenames.</p>
<ul>
<li>Trimmed and quality filtered reads
<ul>
<li><strong>trimmed and filtered reads</strong>. A fastq.gz file containing the reads after removing adapters and low-quality reads. <code>&lt;riboseq_data&gt;/without-adapters/&lt;sample-name&gt;[.&lt;note&gt;].fastq.gz</code></li>
</ul></li>
<li>Reads aligning to ribosomal sequences
<ul>
<li><strong>discarded reads</strong>. A fastq.gz file containing reads which align to the ribosomal index. They are recorded but not used in later processing. <code>&lt;riboseq_data&gt;/with-rrna/&lt;sample-name&gt;[.&lt;note&gt;].fastq.gz</code></li>
</ul></li>
<li>Reads not aligning to ribosomal sequences
<ul>
<li><strong>retained reads</strong>. A fastq.gz file containing reads which do not align to the ribosomal index and are used in further processing. <code>&lt;riboseq_data&gt;/without-rrna/&lt;sample-name&gt;[.&lt;note&gt;].fastq.gz</code></li>
</ul></li>
<li>Aligned reads
<ul>
<li><p><strong>sorted reads aligned to the genome</strong>. A sorted bam file containing all alignments of reads to the genome. <code>&lt;riboseq_data&gt;/without-rrna-mapping/&lt;sample-name&gt;[.&lt;note&gt;]Aligned.sortedByCoord.out.bam</code>. Additionally, <code>&lt;riboseq_data&gt;/without-rrna-mapping/&lt;sample-name&gt;[.&lt;note&gt;].bam</code>, which is a symlink to the <code>Aligned.sortedByCoord.out.bam</code> file.</p></li>
<li><p><strong>aligned reads which map uniquely to the genome</strong>. A sorted bam file containing all alignments of reads to the genome with multimapping reads filtered out. <code>&lt;riboseq_data&gt;/without-rrna-mapping/&lt;sample-name&gt;[.&lt;note&gt;]-unique.bam</code></p></li>
</ul></li>
<li>Metagene profiles
<ul>
<li><p><strong>metagene profiles</strong>. A gzipped csv file containing the metagene profiles for all read lengths which occur in the uniquely-aligning reads. It includes the metagene profile around both the annotated translation initiation site and translation termination site. <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.metagene-profile.csv.gz</code></p></li>
<li><p><strong>periodicity estimations</strong>. A gzipped csv file containing the Bayes factor estimates for all P-site offsets specified by the configuration, as well as information about the number of reads in the respective profile. <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.metagene-periodicity-bayes-factors.csv.gz</code></p></li>
<li><p><strong>estimated P-site offsets</strong>. A gzipped csv file containing the selected P-site offset for each read length. All read lengths are included, even if the estimates do not meet the criteria specified in the configuration file. (The filtering occurs later.) <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.periodic-offsets.csv.gz</code></p></li>
</ul></li>
<li>ORF profiles
<ul>
<li><p><strong>unsmoothed ORF profiles</strong>. A gzipped, sparse <a href="http://math.nist.gov/MatrixMarket/formats.html">matrix market file</a> containing the profiles for all ORFs. <strong>N.B.</strong> The matrix market format uses base-1 indices. <code>&lt;riboseq_data&gt;/orf-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.profiles.mtx.gz</code></p></li>
<li><p><strong>smoothed ORF profiles</strong>. The smoothed profiles are not explicitly stored.</p></li>
</ul></li>
</ul>
<p>Indices are also created for the bam files. STAR creates parameter files in the location <code>riboseq_data</code>/without-rrna-mapping/<code>sample-name</code>_STARgenome.</p>
<h3 id="difference-from-paper">Difference from paper</h3>
<p>The fifth step of creating the base genome profile in the paper is “Everything except the 5’ end of the remaining reads is removed.” This profile is not explicitly constructed in the pipeline. The <code>&lt;sample_name&gt;-unique.bam</code> file already contains the necessary information.</p>
<pre><code>create-orf-profiles test-chrI.fastq.gz c-elegans-test.yaml c-elegans-chrI --num-cpus 2 --tmp /scratch/bmalone/ --keep-intermediate-files</code></pre>
<p><a href="#toc">Back to top</a></p>
<p><a id='predicting-translated-open-reading-frames'></a></p>
<h2 id="predicting-translated-open-reading-frames">Predicting translated open reading frames</h2>
<p>The entire profile creation and translation prediction process can be run automatically using the <code>predict-translated-orfs</code> script. It reads most of the required paths from a <a href="http://www.yaml.org/start.html">YAML</a> configuration file. Additionally, it automatically creates some of the output paths.</p>
<p>The script accepts a <code>--overwrite</code> flag. Unless this is given, then steps for which the output file already exists will be skipped.</p>
<p>It also accepts a <code>--do-not-call</code> flag. If this flag is given, then the commands below will be printed but not executed.</p>
<p><a href="#logging-options">Logging options</a> can also be given to this script.</p>
<p><a href="#parallel-processing-options">Parallel processing options</a> can be given to this script.</p>
<h3 id="configuration-file-keys-3">Configuration file keys</h3>
<p>The following keys are read from the configuration file. Keys with [<code>brackets</code>] are optional.</p>
<ul>
<li><code>riboseq_data</code>. The base output location for all created files</li>
<li>[<code>note</code>]. An optional string which will be added to all filenames. It should not contain spaces or any other special characters.</li>
<li>[<code>models_base</code>]. The base path to the compiled models, the base path to the compiled models. The models specified in the paper are included with the source distribution and compiled/pickled as part of the installation process. They are installed in an operating system-specific location (in particular, <code>user_data_dir</code> from the <a href="https://pypi.python.org/pypi/appdirs">appdirs package</a>. This location is determined during installation and does not normally need to be changed. For development, they may be in some alternative location.</li>
</ul>
<h4 id="reference-genome-options-2">Reference genome options</h4>
<p>These options should be exactly the same as those used in the configuration file used to create the reference indices.</p>
<ul>
<li><code>fasta</code>. The path to the reference genome sequence</li>
<li><code>genome_base_path</code>. The path to the output directory for the transcript fasta and ORFs</li>
<li><code>genome_name</code>. A descriptive name to use for the created files</li>
</ul>
<h4 id="bayes-factor-estimate-options">Bayes factor estimate options</h4>
<ul>
<li><p>[<code>chi_square_only</code>]. If this flag is in the config file with any value, then only the Rp-chi pipeline will be performed; namely, the translation models will not be fit to the data, and the posterior distributions will not be estimated.</p></li>
<li><p>[<code>translation_iterations</code>]. The number of iterations to use for each chain in the MCMC sampling. The first half of the iterations are discarded as burn-in samples. All of the remaining samples are used to estimate the posterior distributions. (That is, we do not use thinning.) default: 200</p></li>
</ul>
<h4 id="selecting-predicted-orfs-options">Selecting predicted ORFs options</h4>
<ul>
<li>[<code>min_bf_mean</code>]. The minimum value for the estimated Bayes factor mean to “predict” that an ORF is translated. This value is used in conjunction with both min_bf_mean and min_bf_likelihood.</li>
</ul>
<p>If <code>max_bf_var</code> is a positive value, then this is taken as a hard threshold on the estimated Bayes factor mean. ORFs must meet both the <code>min_bf_mean</code> and <code>max_bf_var</code> filters to be selected as “translated.”</p>
<p>If <code>min_bf_likelihood</code> is given, then this is taken as the boundary value; that is, an ORF is selected as “translated” if:</p>
<pre><code>   [P(bf &gt; min_bf_mean)] &gt; min_bf_likelihood</code></pre>
<p>If both <code>max_bf_var</code> and <code>min_bf_likelihood</code> are <code>None</code> (null in YAML), then this is taken as a hard threshold on the mean for selecting translated ORFs.</p>
<p>If both <code>max_bf_var</code> and <code>min_bf_likelihood</code> are given, then both filters will be applied and the result will be the intersection.</p>
<p>default: 5</p>
<ul>
<li><p>[<code>max_bf_var</code>]. The maximum value value for the estimated Bayes factor variance to “predict” that an ORF is translated. ORFs must meet both the <code>min_bf_mean</code> and <code>max_bf_var</code> filters to be predicted. See the description for <code>min_bf_mean</code> for more details. default: null (i.e., this filter is not used by default)</p></li>
<li><p>[<code>min_bf_likelihood</code>]. The minimum probability of the BF exceeding <code>min_bf_mean</code> to select an ORF as translated. See the description for <code>min_bf_mean</code> for more details. default: 0.5</p></li>
<li><p>[<code>chisq_significance_level</code>]. For the chi-square test, this value is first Bonferroni corrected based on the number of ORFs which pass the smoothing filters. It is then used as the significance threshold to select translated ORFs. default: 0.01</p></li>
</ul>
<h4 id="shared-mcmc-options-1">Shared MCMC options</h4>
<p>These affect the MCMC both for estimating metagene profile periodicity and ORF translation Bayes factors.</p>
<ul>
<li>[<code>seed</code>]. The random seed for the MCMC sampling. default: 8675309</li>
<li>[<code>chains</code>]. The number of chains to use in the MCMC sampling. default: 2</li>
</ul>
<h3 id="input-files-2">Input files</h3>
<p>This script requires several files created during the previous steps in the pipeline, as well as a few external files.</p>
<ul>
<li>External files
<ul>
<li><strong>genome fasta file</strong>. The genome fasta file. This is the same file used for <code>prepare-rpbp-genome</code>.</li>
<li><strong>orfs</strong>. The ORFs (gzipped bed12+ file) created by <code>prepare-rpbp-genome</code>. It must be located at <code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.genomic-orfs.&lt;orf_note&gt;.bed.gz</code></li>
<li><strong>exons</strong>. The ORF exons (gzipped bed6+ file) created by <code>prepare-rpbp-genome</code>. It must be located at <code>&lt;genome_base_path&gt;/transcript-index/&lt;genome_name&gt;.orfs.&lt;orf_note&gt;.bed.gz</code></li>
<li><strong>models of translation</strong>. Some compiled, pickled Stan model files must be located in the <code>&lt;models_base&gt;/translated</code> folder.</li>
<li><strong>models of lack of translation</strong>. Some compiled, pickled Stan model files must be located in the <code>&lt;models_base&gt;/untranslated</code> folder.</li>
</ul></li>
<li>Metagene profiles
<ul>
<li><p><strong>metagene profiles</strong>. A gzipped csv file containing the metagene profiles for all read lengths which occur in the uniquely-aligning reads. It includes the metagene profile around both the annotated translation initiation site and translation termination site. <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.metagene-profile.csv.gz</code></p></li>
<li><p><strong>periodicity estimations</strong>. A gzipped csv file containing the Bayes factor estimates for all P-site offsets specified by the configuration, as well as information about the number of reads in the respective profile. <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.metagene-periodicity-bayes-factors.csv.gz</code></p></li>
<li><p><strong>estimated P-site offsets</strong>. A gzipped csv file containing the selected P-site offset for each read length. All read lengths are included, even if the estimates do not meet the criteria specified in the configuration file. (The filtering occurs later.) <code>&lt;riboseq_data&gt;/metagene-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.periodic-offsets.csv.gz</code></p></li>
</ul></li>
<li>ORF profiles
<ul>
<li><strong>unsmoothed ORF profiles</strong>. A gzipped, sparse <a href="http://math.nist.gov/MatrixMarket/formats.html">matrix market file</a> containing the profiles for all ORFs. <strong>N.B.</strong> The matrix market format uses base-1 indices. <code>&lt;riboseq_data&gt;/orf-profiles/&lt;sample-name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.profiles.mtx.gz</code></li>
</ul></li>
</ul>
<h3 id="output-files-2">Output files</h3>
<p>If replicates are merged, then these files will be created for each condition. Otherwise, they will be created for each sample (or both if the appropriate options are given). Note that the <code>frac-&lt;fraction&gt;.rw-&lt;reweighting_iterations&gt;</code> part of the file names is omitted for default values, unless specified in the configuration file.</p>
<ul>
<li>Estimates
<ul>
<li><strong>the Bayes factor estimates</strong>. A BED12+ file which contains the estimated values for all ORFs (which pass the thresholds mentioned above). The first 12 columns are valid BED12 entries that are simply copied from the <code>orfs</code> BED file. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.smooth.frac-&lt;fraction&gt;.rw-&lt;reweighting_iterations&gt;.bayes-factors.bed.gz</code></li>
</ul></li>
<li>Predictions
<ul>
<li><strong>Rp-Bp predictions</strong> are made using the methodology described in the paper.
<ul>
<li><p><strong>the ORFs</strong>. A BED12+ file containing the ORFs in the final prediction set. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.smooth.frac-&lt;fraction&gt;.rw-&lt;reweighting_iterations&gt;.predicted-orfs.bed.gz</code></p></li>
<li><p><strong>the DNA sequences</strong>. A fasta file containing the DNA sequences of the predicted ORFs. The fasta header matches the ‘id’ column in the BED files. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.smooth.frac-&lt;fraction&gt;.rw-&lt;reweighting_iterations&gt;.predicted-orfs.dna.gz</code></p></li>
<li><p><strong>the protein sequences</strong>. A fasta file containing the protein sequences of the predicted ORFs. The fasta header matches the ‘id’ column in the BED files. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.smooth.frac-&lt;fraction&gt;.rw-&lt;reweighting_iterations&gt;.predicted-orfs.protein.gz</code></p></li>
</ul></li>
<li><p><strong>Rp-chi predictions</strong> are made using a simple chi-square test.</p>
<ul>
<li><p><strong>the ORFs</strong>. A BED12+ file containing the ORFs in the final prediction set. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.chisq.predicted-orfs.bed.gz</code></p></li>
<li><p><strong>the DNA sequences</strong>. A fasta file containing the DNA sequences of the predicted ORFs. The fasta header matches the ‘id’ column in the BED files. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.chisq.predicted-orfs.dna.gz</code></p></li>
<li><p><strong>the protein sequences</strong>. A fasta file containing the protein sequences of the predicted ORFs. The fasta header matches the ‘id’ column in the BED files. <code>&lt;riboseq_data&gt;/orf-predictions/&lt;sample_name&gt;[.&lt;note&gt;]-unique.length-&lt;lengths&gt;.offset-&lt;offsets&gt;.chisq.predicted-orfs.protein.gz</code></p></li>
</ul></li>
</ul></li>
</ul>
<p>Furthermore, there are “unfiltered” and “filtered” versions of the files. The “filtered” versions result from performing the filtering described in the paper (taking the longest predicted ORF for each stop codon, and then selecting the ORF with the highest expected Bayes factor among each group of overlapping ORFs). The “unfiltered” version contains all predictions.</p>
<pre><code>predict-translated-orfs c-elegans-test.yaml c-elegans-chrI --num-cpus 2  --tmp /scratch/bmalone/</code></pre>
<p><a href="#toc">Back to top</a></p>
<p><a id='using-existing-alignment-files'></a></p>
<h2 id="using-existing-alignment-files">Using existing alignment files</h2>
<p>While we found the <code>flexbar</code>-<code>bowtie2</code>-<code>STAR</code> pipeline to be effective for processing and aligning reads, the probabilistic models of Rp-Bp do not depend on any particular properties of these programs. Thus, it is possible to use any alignment files for the pipeline. Essentially, the alignment (bam) files need to be named as expected by the pipeline and placed in the correct folders.</p>
<p>Please see the <a href="custom-alignment-files.html">custom alignment files</a> documentation for more details.</p>
<p><a href="#toc">Back to top</a></p>
<p><a id='logging-options'></a></p>
<h2 id="logging-options">Logging options</h2>
<p>All of the driver scripts mentioned above, and many of the internal scripts as well, allow detailed specification of logging options on the command line. Interally, logging is handled using the standard python logging system. The following options are allowed.</p>
<p>The allowed logging levels are: <code>NOTSET</code>, <code>DEBUG</code>, <code>INFO</code>, <code>WARNING</code>, <code>ERROR</code>, <code>CRITICAL</code>. Most of the scripts output possibly-useful information at the <code>INFO</code> level; some scripts also output <code>DEBUG</code> messages, but as the name indicates, they are typically only useful for tracking down problems which do not cause actual errors.</p>
<ul>
<li><p><code>--log-file</code>. This option specifies a file to which logging statements will be written (in addition to stdout and stderr, if specified). default: False</p></li>
<li><p><code>--log-stdout</code>. If this flag is present, then logging statements will be written to stdout (in addition to a file and stderr, if specified). default: False</p></li>
<li><p><code>--log-stderr</code>. If this flag is present, then logging statements will be written to stderr (in addition to a file and stdout, if specified). default: True</p></li>
<li><p><code>--logging-level</code>. If this value is specified, then it will be used for all logs. default: WARNING</p></li>
<li><p><code>--file-logging-level</code>. The logging level to be used for the log file, if specified. This option overrides <code>--logging-level</code>. default: NOTSET</p></li>
<li><p><code>--stdout-logging-level</code>. The logging level to be used for the stdout log, if specified. This option overrides <code>--logging-level</code>. default: NOTSET</p></li>
<li><p><code>--stderr-logging-level</code>. The logging level to be used for the stderr log, if specified. This option overrides <code>--logging-level</code>. default: NOTSET</p></li>
</ul>
<p><a href="#toc">Back to top</a></p>
<p><a id='parallel-processing-options'></a></p>
<h2 id="parallel-processing-options">Parallel processing options</h2>
<p>All of the scripts accept options for running code in parallel. Furthermore, the scripts are designed to seamlessly integrate with the <a href="http://slurm.schedmd.com/">SLURM</a> scheduler. The following options are allowed. SLURM-specific options are specified by [<code>brackets</code>].</p>
<ul>
<li><p><code>--num-cpus</code>. The number of CPUs to use. The definition of a “CPU” varies somewhat among the programs. For example, for STAR, these are actually threads. For many of the python scripts, this number is translated into the number of processes to spawn. None of the code parallelizes across machines, so the value should not be greater than the number of cores on the machine on which the programs are executed. When used with SLURM, this will be translated into an sbatch request like: “–ntasks 1 –cpus-per-task &lt;num-cpus&gt;”. default: 1</p></li>
<li><p><code>--mem</code>. For STAR genome indexing, the amount of RAM to request. The rest of the programs do not use this value. When used with SLURM, this will be translated into an sbatch request like: “–mem=&lt;mem&gt;”. default: 10G</p></li>
<li><p><code>--do-not-call</code>. If this flag is present, then the commands will not be executed (but will be printed). This can be useful to ensure paths in configuration files are correct.</p></li>
<li><p>[<code>--use-slurm</code>]. If this flag is present, then the commands will be submitted to SLURM via sbatch. default: By default, each command is executed sequentially within the current terminal.</p></li>
<li><p>[<code>--time</code>]. The amount of time to request. This will be translated into an sbatch request like: “–time &lt;time&gt;. default: 0-05:59</p></li>
<li><p>[<code>--partitions</code>]. The partitions to request. This will be translated into an sbatch request like: “-p &lt;partitions&gt;”. default: general (N.B. This value should be a comma-separated list with no spaces, for example: <code>--partitions general,long</code>)</p></li>
<li><p>[<code>--no-output</code>]. If this flag is present, stdout will be redirected to /dev/null. This will be translated into an sbatch request like: “–output=/dev/null”. default: If the flag is not present, then stdout will be directed to a log file with the job number. This corresponds to “–output=slurm-%J.out” in the sbatch call.</p></li>
<li><p>[<code>--no-error</code>]. If this flag is present, stderr will be redirected to /dev/null. This will be translated into an sbatch request like: “–error=/dev/null”. default: If the flag is not present, then stderr will be directed to a log file with the job number. This corresponds to “–error=slurm-%J.err” in the sbatch call.</p></li>
<li><p>[<code>--stdout-file</code>]. If this is present and the –no-output flag is not given, then stdout will be directed to this file. This corresponds to “–output=<code>stdout-file</code>” in the sbatch call. default: slurm-%J.out</p></li>
<li><p>[<code>--stderr-file</code>]. If this is present and the –no-error flag is not given, then stderr will be directed to this file. This corresponds to “–error=<code>stderr-file</code>” in the sbatch call. default: slurm-%J.err</p></li>
<li><p>[<code>--mail-type</code>]. When to send an email notifcation of the job status. See official documentation for a description of the values. If a mail-user is not specified, this will revert to ‘None’. Defaut: FAIL TIME_LIMIT</p></li>
<li><p>[<code>--mail-user</code>]. To whom an email will be sent, in accordance with mail-type. default: None</p></li>
</ul>
<pre><code># This will submit the prepare-genome script to SLURM as a single job. That job
# will request 10 CPUs and 100G of RAM.

prepare-rpbp-genome WBcel235.79.chrI.yaml --num-cpus 10 --mem 100G --overwrite --logging-level INFO --use-slurm --tmp /scratch/bmalone/</code></pre>
<pre><code># This will submit each sample as a separate job to SLURM. Each submitted job will request 10 cpus
# and 100G of RAM. For example, if c-elegans-test.yaml specifies 5 samples in the riboseq_samples
# value, then 5 jobs will be submitted to SLURM, one for each sample.

# If the --merge-replicates flag is given, then all of the profiles are first created. Then, the
# combined profiles are created in accordance with riboseq_biological_replicates from the config
# file. Finally, the last phase of the pipeline (predict-translated-orfs) is called.

# If any step fails, then later phases will not be started.

run-all-rpbp-instances c-elegans-test.yaml --tmp /scratch/bmalone/ --num-cpus 10 --mem 100G --merge-replicates --use-slurm --logging-level INFO</code></pre>
<p><a href="#toc">Back to top</a></p>
</body>
</html>
